### PR TITLE
Rag augment

### PR DESCRIPTION
**Pull request by kemmaliu**
integrated RAG with VerMCTS for Dafny context using the Devin chunked dafny manual

### CODE REVIEW

File: default_prompts.py [hunk 1]
```diff
@@ -74,9 +74,11 @@
     5,
     25,
     NO_CHECK_PROOF, NO_CHECK_CHEAT,
-    ALL_LANGS,
+    ['Dafny'],
```
Comment (namin) [inline default_prompts.py:77]: Why are you changing this to be Dafny only? Also why is the test in Python?

File: python.py [hunk 2]
```diff
@@ -47,10 +47,13 @@ def calculateScoreHelper(msg: str) -> (Optional[float], Optional[str]):
...
-    first = log[log.rindex(marker) + len(marker):]
-    num_line_first = int(first[0 : find_first_index(first, '\n', ',')])
-    if filter_code(msg).strip() != v and num_line_first >= v.count("\n"):
-        return None, None
+    try:
+        first = log[log.rindex(marker) + len(marker):]
+        num_line_first = int(first[0 : find_first_index(first, '\n', ',')])
+        if filter_code(msg).strip() != v and num_line_first >= v.count("\n"):
+            return None, None
+    except:
```
Comment (namin) [inline python.py:55]: Can you be more specific with the error you're catching? and why?

File: python.py [hunk 3]
```diff
@@ -63,11 +66,11 @@ def runUnittests(msg: str) -> (Optional[float], Optional[str]):
         return 1.0, None
     log = r["log"]
     print(log)
-    marker = "ex.py\", line "
+    '''marker = "ex.py\", line "
```
Comment (namin) [inline python.py:69]: What is this?

File: run_intermediate_expansion.py [hunk 4]
```diff
@@ -121,4 +121,4 @@ def main(mins_timeout=None, prompt=prompt):
 
 
 if __name__ == "__main__":
-    main()
+    main()
```
Comment (namin) [inline run_intermediate_expansion.py:124]: no need to change that file?

File: run_intermediate_expansion_mod.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
+from montecarlo.node import Node
```
Comment (namin) [inline run_intermediate_expansion_mod.py:1]: _mod is a bad name? should it be _rag?

File: run_intermediate_expansion_mod.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+def main(mins_timeout=None, prompt=prompt):
+    # embed model
+    Settings.embed_model = HuggingFaceEmbedding(
+        model_name="BAAI/bge-small-en-v1.5"
+    )
+    # rebuild storage context
+    storage_context = StorageContext.from_defaults(persist_dir="./storage")
+    # load index
+    index = load_index_from_storage(storage_context)
```
Comment (namin) [inline run_intermediate_expansion_mod.py:113]: how to create the storage?

File: setup_index.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
+#set up the index ONCE
```
Comment (namin) [inline setup_index.py:1]: Put a space after #

File: setup_index.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+from llama_index.core import VectorStoreIndex, SimpleDirectoryReader, Settings
+from llama_index.embeddings.huggingface import HuggingFaceEmbedding
+
+
+
+Settings.embed_model = HuggingFaceEmbedding(
```
Comment (namin) [inline setup_index.py:7]: you're doing this Settings.embed_model in three different places, maybe you should have a file that just does RAG settings?

**Review by namin — CHANGES_REQUESTED** [review]

File: default_prompts.py (hunk 1 above)
Comment (kemmaliu) [reply default_prompts.py]: changed it back to ALL_LANGS. As for the Python test, I believe that change was from when we were working on Python Testing with Simon

File: python.py (hunk 2 above)
Comment (kemmaliu) [reply python.py]: I ended up changing the score function back to what it was before hand!

File: default_prompts.py [hunk 9]
```diff
@@ -76,7 +76,9 @@
     NO_CHECK_PROOF, NO_CHECK_CHEAT,
     ALL_LANGS,
     None,
-    None
+    {
+        "def fib": "test(fib(3) == 1)"
```
Comment (namin) [inline default_prompts.py:80]: just remove this test

File: call_rag.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        if not prompt.startswith("<|system|>\n"):
+            prompt = "<|system|>\n</s>\n" + prompt
+
+        # add final assistant prompt
+        prompt = prompt + "<|assistant|>\n"
+
+        return prompt
+
+
+    Settings.llm = HuggingFaceLLM(
```
Comment (namin) [inline call_rag.py:37]: I think you want to do this setup outside the call?

**Review by namin — CHANGES_REQUESTED** [review]
almost there

**Review by namin — APPROVED** [review]

**Pull request merged**